### PR TITLE
Implement Handler versions

### PR DESCRIPTION
**Pull request by flux627**
This PR:

- Implements multiple, versioned sets of updaters/effects (called `HandlerVersion`s)
- Implements the ability to switch between `HandlerVersion`s by returning a value from any Updater
- Warns user for various misuse scenarios
- Refactor interfaces
- Update some naming conventions

Would like constructive feedback on the API, open to improvements!

### CODE REVIEW

File: src/AbstractActionHandler.ts [hunk 1]
```diff
@@ -33,8 +35,10 @@ export abstract class AbstractActionHandler {
       console.info(`Rolling back ${rollbackCount} blocks to block ${rollbackBlockNumber}...`)
       await this.rollbackTo(rollbackBlockNumber)
       await this.refreshIndexState()
-    } else if (!this.lastProcessedBlockHash && this.lastProcessedBlockNumber === 0) {
+      await this.refreshHandlerVersionState()
+    } else if (this.lastProcessedBlockNumber === 0 && this.lastProcessedBlockHash === "") {
```
Comment (jeffreyssmith2nd) [inline src/AbstractActionHandler.ts]: I'm curious, why the change here?

File: src/AbstractActionHandler.ts [hunk 2]
```diff
@@ -90,36 +98,57 @@ export abstract class AbstractActionHandler {
   /**
    * Process actions against deterministically accumulating updater functions.
    */
-  protected async runUpdaters(
+  protected async applyUpdaters(
```
Comment (jeffreyssmith2nd) [inline src/AbstractActionHandler.ts]: This could benefit from some method extraction. It may just be the logging, but its getting complicated to read/follow

File: src/testHelpers/TestActionHandler.ts [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
-import { Block, IndexState } from "../interfaces"
+import { Action, Block, IndexState } from "../interfaces"
 
 export class TestActionHandler extends AbstractActionHandler {
   public state: any = {
-    indexState: { blockNumber: 0, blockHash: "" },
+    indexState: { blockNumber: 0, blockHash: "", isReplay: false, handlerVersion: "v1" },
   }
 
+  get _handlerVersionName() { return this.handlerVersionName }
```
Comment (jeffreyssmith2nd) [inline src/testHelpers/TestActionHandler.ts]: What is this syntax doing?

Comment (jeffreyssmith2nd) [reply src/testHelpers/TestActionHandler.ts]: `get` specifically

Comment (flux627) [reply src/testHelpers/TestActionHandler.ts]: It means that I can
```
handlerInstance._handlerVersionName
```
rather than 
```
handlerInstance._handlerVersionName
```

File: src/AbstractActionHandler.ts (hunk 1 above)
Comment (flux627) [reply src/AbstractActionHandler.ts]: More explicit- the initial state of `this.lastProcessedBlockHash` is an empty string, so when reading the code it's easy to see that that's what I'm checking for.

File: src/AbstractActionHandler.ts (hunk 2 above)
Comment (flux627) [reply src/AbstractActionHandler.ts]: Extracting the warnings into their own functions did help readability a lot.
